File: sorting-searching/sorting-algorithms/SIMD-Sorting/simd-sorting.py
BLOCK_SIZE = 4

def merge_two_sorted_arrays(left, right):
    result = []

    i = 0
    j = 0

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    while i < len(left):
        result.append(left[i])
        i += 1

    while j < len(right):
        result.append(right[j])
        j += 1

    return result


def simd_style_sort(arr):
    sorted_blocks = []

    for i in range(0, len(arr), BLOCK_SIZE):
        block = arr[i:i + BLOCK_SIZE]
        block.sort()
        sorted_blocks.append(block)

    while len(sorted_blocks) > 1:
        merged_blocks = []

        for i in range(0, len(sorted_blocks), 2):
            if i + 1 < len(sorted_blocks):
                merged_blocks.append(
                    merge_two_sorted_arrays(sorted_blocks[i], sorted_blocks[i + 1])
                )
            else:
                merged_blocks.append(sorted_blocks[i])

        sorted_blocks = merged_blocks

    return sorted_blocks[0] if sorted_blocks else []


arr = [8, 3, 7, 4, 9, 2, 6, 1, 10, 5, 12, 11]

print("Original array:", arr)

sorted_array = simd_style_sort(arr)

print("Sorted array:", sorted_array)

File: sorting-searching/sorting-algorithms/SIMD-Sorting/simd-sorting.js
const BLOCK_SIZE = 4;

function mergeTwoSortedArrays(left, right) {
  const result = [];

  let i = 0;
  let j = 0;

  while (i < left.length && j < right.length) {
    if (left[i] <= right[j]) {
      result.push(left[i]);
      i++;
    } else {
      result.push(right[j]);
      j++;
    }
  }

  while (i < left.length) {
    result.push(left[i]);
    i++;
  }

  while (j < right.length) {
    result.push(right[j]);
    j++;
  }

  return result;
}

function simdStyleSort(arr) {
  const sortedBlocks = [];

  for (let i = 0; i < arr.length; i += BLOCK_SIZE) {
    const block = arr.slice(i, i + BLOCK_SIZE);

    block.sort((a, b) => a - b);

    sortedBlocks.push(block);
  }

  while (sortedBlocks.length > 1) {
    const mergedBlocks = [];

    for (let i = 0; i < sortedBlocks.length; i += 2) {
      if (i + 1 < sortedBlocks.length) {
        mergedBlocks.push(
          mergeTwoSortedArrays(sortedBlocks[i], sortedBlocks[i + 1])
        );
      } else {
        mergedBlocks.push(sortedBlocks[i]);
      }
    }

    sortedBlocks.length = 0;
    sortedBlocks.push(...mergedBlocks);
  }

  return sortedBlocks.length > 0 ? sortedBlocks[0] : [];
}

const arr = [8, 3, 7, 4, 9, 2, 6, 1, 10, 5, 12, 11];

console.log("Original array:", arr);

const sortedArray = simdStyleSort(arr);

console.log("Sorted array:", sortedArray);
